### PR TITLE
conn_pool: track streams across the pool

### PR DESCRIPTION
**Pull request by alyssawilk**
Tracking active, pending, and available capacity for each thread local cluster, for eventual use in cluster-wide prefetch

Risk Level: low
Testing: new unit tests
Docs Changes: n/a
Release Notes: n/a

### CODE REVIEW

**Comment by alyssawilk** [comment]
I had originally planned to land this with the use in source/common/upstream/cluster_manager_impl.cc but it was a large and fragile enough change I think I'd prefer them to be separate (unless you prefer otherwise)

I really don't like how fragile the stream tracking is, but there's a weak link between the codec's streams and the connection pool's streams with different lifetimes, which I'm not sure how to address better than I have here.

File: source/common/upstream/load_balancer_impl.cc [hunk 1]
```diff
@@ -232,6 +233,11 @@ void LoadBalancerBase::recalculatePerPriorityState(uint32_t priority,
                                 per_priority_load.healthy_priority_load_.get().end(), 0) +
                     std::accumulate(per_priority_load.degraded_priority_load_.get().begin(),
                                     per_priority_load.degraded_priority_load_.get().end(), 0));
+
+  total_healthy_hosts = 0;
```
Comment (antoniovicente) [inline source/common/upstream/load_balancer_impl.cc:237]: This function contains an early return on line 195; total_healthy_hosts is not initialized on early return.

**Review by antoniovicente — COMMENTED** [review]
> I had originally planned to land this with the use in source/common/upstream/cluster_manager_impl.cc but it was a large and fragile enough change I think I'd prefer them to be separate (unless you prefer otherwise)
> 
> I really don't like how fragile the stream tracking is, but there's a weak link between the codec's streams and the connection pool's streams with different lifetimes, which I'm not sure how to address better than I have here.

I did a review pass.  The main thing that jumped at me that concurrent_stream_limit can be set to uint64_t max which may cause some of the tracking counters to behave erratically.

File: include/envoy/upstream/cluster_manager.h [hunk 2]
```diff
@@ -73,6 +73,34 @@ using ClusterUpdateCallbacksHandlePtr = std::unique_ptr<ClusterUpdateCallbacksHa
...
+    ASSERT(pending_streams_ == 0);
+    ASSERT(active_streams_ == 0);
+    ASSERT(connecting_capacity_ == 0);
+  }
+  void checkAndDecrement(uint32_t& value, uint32_t& delta) {
+    ASSERT(value - delta <= value);
+    value -= delta;
+  }
+
+  void incrPendingStreams(uint32_t delta) { pending_streams_ += delta; }
```
[the remark this answers is not in the document]
Comment (alyssawilk) [reply include/envoy/upstream/cluster_manager.h:88]: and the decrement asserts, which caught a lot of issues earlier on.

File: source/common/conn_pool/conn_pool_base.cc [hunk 3]
```diff
@@ -107,6 +108,8 @@ bool ConnPoolImplBase::tryCreateNewConnection(float global_prefetch_ratio) {
     ASSERT(std::numeric_limits<uint64_t>::max() - connecting_stream_capacity_ >=
            client->effectiveConcurrentStreamLimit());
     ASSERT(client->real_host_description_);
+    // Increase the connecting capacity to reflect the streams this connection can serve.
+    state_.incrConnectingCapacity(client->effectiveConcurrentStreamLimit());
```
[the remark this answers is not in the document]
Comment (alyssawilk) [reply source/common/conn_pool/conn_pool_base.cc:112]: Yeah, in practice these are set by config, and 

  google.protobuf.UInt32Value max_concurrent_streams = 2
      [(validate.rules).uint32 = {lte: 2147483647 gte: 1}];

so they're functionally capped at uint32.  Would you prefer I make all the base class types uint32_t, and the new types uint64 for overflow safety?

[the remark this answers is not in the document]
Comment (alyssawilk) [reply source/common/conn_pool/conn_pool_base.cc:112]: ping?

File: include/envoy/upstream/cluster_manager.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (antoniovicente) [reply include/envoy/upstream/cluster_manager.h:88]: Can we ASSERT that we don't overflow on increment also?

File: source/common/conn_pool/conn_pool_base.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (antoniovicente) [reply source/common/conn_pool/conn_pool_base.cc:112]: Sorry, I didn't realize that you had replied to comments until now.  I should have tagged the PR with waiting:any instead of waiting.

What if someone does not set config and ends up with the default concurrent_stream_limit of 0?
I did some tracing through the H2 implementation and found that initializeAndValidateOptions does translate a config value of 0 to 2147483647 as documented.  It seems to me that concurrent_stream_limit should never be 0 when it reaches this constructor.

Still, connecting_capacity_ is an uint32_t.  If there are more than 2 endpoints in the cluster, I think this will overflow when using the H2 default of 2**31-1 since I think these stats are per cluster, not per host/endpoint.

**Comment by alyssawilk** [comment]
I think I've addressed the overflow concerns.  PTAL?

**Comment by alyssawilk** [comment]
ping :-)

**Review by antoniovicente — APPROVED** [review]

**Review by mattklein123 — COMMENTED** [review]
At a high level this LGTM. 

I do have one question though: are we duplicating information that we already have in metrics? There are two parts to this question:
1) On one hand, we are trying to not drive functionality through stats because it leads to bugs when stats get excluded.
2) On the other hand we might be duplicating accounting logic and then not have visibility into certain metrics.

To account for (2) we have discussed potentially allowing stats to be marked as not excludable if we know they are used in critical logic. So my question is did you consider using stats for this and decide not to because of potential exclusion bugs? cc @jmarantz 

/wait-any

**Comment by jmarantz** [comment]
Having some stats be force-included independent of the disallow patterns makes sense to me. Is there an open bug for that?

**Comment by mattklein123** [comment]
> Having some stats be force-included independent of the disallow patterns makes sense to me. Is there an open bug for that?

I don't think it's tracked yet, but it does keep coming up and it seems like we might want that capability.

**Comment by alyssawilk** [comment]
There may be overlap with stats, but I don't think we can use stats because prefetch data needs to be per-worker since the connections are only available locally (until/if we have cross-thread pools)

**Comment by mattklein123** [comment]
> There may be overlap with stats, but I don't think we can use stats because prefetch data needs to be per-worker since the connections are only available locally (until/if we have cross-thread pools)

OK that makes sense. I will take another quick pass today.

File: include/envoy/upstream/cluster_manager.h [hunk 4]
```diff
@@ -73,6 +73,44 @@ using ClusterUpdateCallbacksHandlePtr = std::unique_ptr<ClusterUpdateCallbacksHa
...
+  void incrActiveStreams(uint32_t delta) { checkAndIncrement<uint32_t>(active_streams_, delta); }
+  void decrActiveStreams(uint32_t delta) { checkAndDecrement<uint32_t>(active_streams_, delta); }
+
+  // Tracks the number of pending streams for this ClusterManager.
+  uint32_t pending_streams_{};
+  // Tracks the number of active streams for this ClusterManager.
+  uint32_t active_streams_{};
+  // Tracks the stream capacity if all connecting connections were connected but
+  // excluding streams which are in use.
+  uint64_t connecting_capacity_{};
```
Comment (mattklein123) [inline include/envoy/upstream/cluster_manager.h:111]: nit: can you name this `connecting_stream_capacity_` or `connection_stream_capacity_` and similar for the accessor methods? I was confused when I first started to look at the code below.

File: include/envoy/upstream/cluster_manager.h [hunk 5]
```diff
@@ -73,6 +73,44 @@ using ClusterUpdateCallbacksHandlePtr = std::unique_ptr<ClusterUpdateCallbacksHa
...
+  }
+  void incrActiveStreams(uint32_t delta) { checkAndIncrement<uint32_t>(active_streams_, delta); }
+  void decrActiveStreams(uint32_t delta) { checkAndDecrement<uint32_t>(active_streams_, delta); }
+
+  // Tracks the number of pending streams for this ClusterManager.
+  uint32_t pending_streams_{};
+  // Tracks the number of active streams for this ClusterManager.
+  uint32_t active_streams_{};
+  // Tracks the stream capacity if all connecting connections were connected but
+  // excluding streams which are in use.
```
Comment (mattklein123) [inline include/envoy/upstream/cluster_manager.h:110]: nit: this is a bit hard to parse. Can you clarify a bit? This is basically: total theoretical capacity  - pending - active?

File: source/common/upstream/cluster_manager_impl.cc [hunk 6]
```diff
@@ -841,8 +841,13 @@ ThreadLocalCluster* ClusterManagerImpl::get(absl::string_view cluster) {
 }
 
 void ClusterManagerImpl::maybePrefetch(
-    ThreadLocalClusterManagerImpl::ClusterEntryPtr& cluster_entry,
+    ThreadLocalClusterManagerImpl::ClusterEntryPtr& cluster_entry, const ClusterConnectivityState&,
```
Comment (mattklein123) [inline source/common/upstream/cluster_manager_impl.cc:844]: It might be better to split this part of the change out from the accounting work but up to you.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
At a high level looks good with some naming nits. This code is all tricky so IMO it would be better to split the accounting changes out from the other changes which seem unrelated, then we can review the integration separately? Up to you.

/wait

File: source/common/upstream/load_balancer_impl.h [hunk 7]
```diff
@@ -154,6 +155,8 @@ class LoadBalancerBase : public LoadBalancer {
   DegradedAvailability per_priority_degraded_;
   // Levels which are in panic
   std::vector<bool> per_priority_panic_;
+  // The total count of healthy hosts across all priority levels.
+  uint32_t total_healthy_hosts_;
```
Comment (mattklein123) [inline source/common/upstream/load_balancer_impl.h:159]: I think it would be better to revert all of this also as unrelated. If not reverted, does this need to be zero initialized somewhere?

**Review by mattklein123 — COMMENTED** [review]
Thanks LGTM with one other potential question/comment.

/wait-any

**Review by mattklein123 — APPROVED** [review]
Nice, thanks.
